web/tests/lynxkite.ts

```diff
@@ -0,0 +1,1615 @@
+    });
+
+    const total = histogram.locator('#histogram-total');
+    protractor.promise.all([total.getText(), res]).then(function (results) {
+      const totalValue = results[0].match(/histogram total: ([0-9,]+)/)[1];
+      const values = results[1];
+      const total = parseInt(totalValue.replace(/,/g, ''));
+      let sum = 0;
+      for (let j = 0; j < values.length; j++) {
+        sum += values[j].value;
+      }
+      expect(total).toEqual(sum);
+    });
+    this.popoff();
+    return res;
+  }
+
+  async visualizeAs(visualization) {
+    this.popup()
+      .locator('#visualize-as-' + visualization)
+      .click();
+    testLib.expectElement(this.visualizedAs(visualization));
+    this.popoff();
+  }
+
+  async visualizedAs(visualization) {
+    return this.element.locator('#visualized-as-' + visualization);
+  }
+
+  async doNotVisualizeAs(visualization) {
+    this.popup()
+      .locator('#visualize-as-' + visualization)
+      .click();
+    testLib.expectNotElement(this.visualizedAs(visualization));
+    this.popoff();
+  }
+
+  async clickMenu(id) {
+    this.popup()
+      .locator('#' + id)
+      .click();
+    this.popoff();
+  }
+}
+
+export class Workspace {
+  readonly page: Page;
+  readonly main: Locator;
+  readonly selector: Locator;
+  readonly board: Locator;
+  constructor(page: Page) {
+    this.page = page;
+    this.main = page.locator('#workspace-entry-point');
+    this.selector = page.locator('.operation-selector');
+    this.board = page.locator('#workspace-drawing-board');
+  }
+
+  async expectCurrentWorkspaceIs(name) {
+    await expect(this.main.locator('#workspace-name')).toHaveText(name);
+    // TODO: check that workspace is error-free
+  }
+
+  async close() {
+    await this.main.locator('#close-workspace').click();
+  }
+
+  async openOperation(name) {
+    this.selector.element(by.id('operation-search')).click();
+    safeSendKeys(this.selector.element(by.id('filter')), name + K.ENTER);
+    return this.selector.locator('operation-selector-entry').get(0);
+  }
+
+  async closeOperationSelector() {
+    this.selector.element(by.id('operation-search')).click();
+  }
+
+  async closeLastPopup() {
+    browser.actions().sendKeys(K.ESCAPE).perform();
+  }
+
+  async duplicate() {
+    browser.actions().sendKeys(K.chord(CTRL, 'c')).sendKeys(K.chord(CTRL, 'v')).perform();
+  }
+
+  async addBoxFromSelector(boxName) {
+    browser
+      .actions()
+      .sendKeys('/' + boxName + K.ENTER + K.ESCAPE)
+      .perform();
+  }
+
+  async addBox(boxData) {
+    const id = boxData.id;
+    const after = boxData.after;
+    const inputs = boxData.inputs;
+    const params = boxData.params;
+    await this.page.evaluate(boxData => {
+      $(document.querySelector('#workspace-drawing-board')).scope().workspace.addBox(
+        boxData.name,
+        { logicalX: boxData.x, logicalY: boxData.y },
+        { boxId: boxData.id });
+    }, boxData);
+    // Wait for the backend to save this box.
```

Comment:
Alternatively we could make it so that it's okay to click a box before it's saved. I've filed https://github.com/lynxkite/lynxkite/issues/308.